contracts/PeeranhaToken.sol

```diff
@@ -53,54 +92,105 @@ contract PeeranhaToken is ERC20Upgradeable, ERC20PausableUpgradeable, ERC20Cappe
+      poolTokens[period] = poolToken;
+    }
     uint256 userReward = getUserReward(weekReward, user, period, poolToken);
 
-    require(userReward != 0, "No reward for you in this period");
+    require(userReward != 0, "no_reward");
 
     emit GetReward(user, period);
-    _mint(user, userReward * FRACTION);
+    _mint(user, userReward);
+  }
+
+  function setStake(address user, uint256 stakeTokens) external {
+    require(stakeTokens <= balanceOf(user), "wrong_stack");
+    uint16 nextPeriod = RewardLib.getPeriod(CommonLib.getTimestamp()) + 1;
+    TokenLib.StakeUserContainer storage stakeUserContainer = userPeriodStake.userPeriodStake[user];
+
+    TokenLib.StakeTotal storage stakeTotal = stakeTotalContainer.stakeTotals[nextPeriod];
+    if (stakeTotalContainer.stakeChangePeriods.length != 0 && stakeTotalContainer.stakeChangePeriods[stakeTotalContainer.stakeChangePeriods.length - 1] != nextPeriod) {
+      uint16 previousPeriod = stakeTotalContainer.stakeChangePeriods[stakeTotalContainer.stakeChangePeriods.length - 1];
+      stakeTotal.totalStakedAmount = stakeTotalContainer.stakeTotals[previousPeriod].totalStakedAmount;
+      stakeTotal.stakingUsersCount = stakeTotalContainer.stakeTotals[previousPeriod].stakingUsersCount;
+      stakeTotalContainer.stakeChangePeriods.push(nextPeriod);
+    } else if (stakeTotalContainer.stakeChangePeriods.length == 0) {
+      stakeTotalContainer.stakeChangePeriods.push(nextPeriod);
+    }
+
+    if (stakeUserContainer.stakeChangePeriods.length != 0) {
+      uint16 lastStakePeriod = stakeUserContainer.stakeChangePeriods[stakeUserContainer.stakeChangePeriods.length - 1];
+      if (lastStakePeriod != nextPeriod) {
+        stakeUserContainer.stakeChangePeriods.push(nextPeriod);
+      }
+
+      if (stakeUserContainer.userStake[lastStakePeriod].totalStakedAmount > stakeTokens) {
+        stakeTotal.totalStakedAmount += stakeUserContainer.userStake[lastStakePeriod].totalStakedAmount - stakeTokens;
+        if (stakeTokens == 0) stakeTotal.stakingUsersCount--;
+        stakeUserContainer.userStake[nextPeriod].changedStake = stakeUserContainer.userStake[lastStakePeriod].totalStakedAmount - stakeTokens;
+
+      } else {
+        stakeTotal.totalStakedAmount += stakeTokens - stakeUserContainer.userStake[lastStakePeriod].totalStakedAmount;
+        if (stakeUserContainer.userStake[lastStakePeriod].totalStakedAmount == 0) stakeTotal.stakingUsersCount++;
+      }
+      stakeUserContainer.userStake[nextPeriod].totalStakedAmount = stakeTokens;
+
+    } else {
+      stakeUserContainer.stakeChangePeriods.push(nextPeriod);
+      stakeUserContainer.userStake[nextPeriod].totalStakedAmount = stakeTokens;
+
+      stakeTotal.totalStakedAmount += stakeTokens;
+      stakeTotal.stakingUsersCount++;
+    }
   }
 
   function getPool(RewardLib.WeekReward memory weekReward, uint16 period) private view returns(uint256) {
     uint256 poolToken = poolTokens[period];
```

Comment:
Let's remove caching in `poolTokens`. I don't think it saves any significant amount of gas.